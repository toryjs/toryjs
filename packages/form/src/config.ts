const interleave = (strings: TemplateStringsArray, ...interpolations: any[]) =>
  strings.reduce(
    (final, str, i) => final + str + (interpolations[i] === undefined ? '' : interpolations[i]),
    ''
  );

type Config = {
  linkElement: any;
  linkElementTarget: (target: string) => any;
  setDirty?(isDirty: boolean): void;
  i18n(strings: TemplateStringsArray, ...interpolations: any[]): string;
};

export const config: Config = {
  i18n: interleave,
  linkElement: 'a',
  linkElementTarget: (target: string) => (target ? { href: target } : undefined)
};
