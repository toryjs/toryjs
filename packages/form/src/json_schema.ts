import { Option } from './form_definition';

export type JSONSchema7Version =
  | 'http://json-schema.org/schema#'
  | 'http://json-schema.org/hyper-schema#'
  | 'http://json-schema.org/draft-07/schema#'
  | 'http://json-schema.org/draft-07/hyper-schema#';

export type JSONSchema7TypeName =
  | 'string'
  | 'number'
  | 'integer'
  | 'boolean'
  | 'object'
  | 'array'
  | 'null';
export type JSONSchema7Type = JSONSchema7Array[] | boolean | number | null | object | string;

// Workaround for infinite type recursion
// https://github.com/Microsoft/TypeScript/issues/3496#issuecomment-128553540
// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface JSONSchema7Array extends Array<JSONSchema7Type> {}

export type JSONSchemaType = JSONSchema7Type;

export type ErrorMessages = {
  [index: string]: ErrorMessage;
};

export type ErrorMessage = string | ErrorMessages;

export class JSONSchemaBase {
  /* =========================================================
      OVERRIDEN
     ======================================================== */

  expression?: string;
  validationExpression?: string;
  validationGroup?: string;
  share?: boolean;

  /**
   * @see https://tools.ietf.org/html/draft-handrews-json-schema-validation-01#section-6.1
   */
  type?: JSONSchema7TypeName;
  enum?: JSONSchema7Type[];
  const?: JSONSchema7Type;

  /* =========================================================
      ORIGINAL
     ======================================================== */

  $id?: string;
  $ref?: string;
  $schema?: JSONSchema7Version;
  $comment?: string;
  $import?: string;
  $enum?: Option[];

  /**
   * @see https://tools.ietf.org/html/draft-handrews-json-schema-validation-01#section-6.2
   */
  multipleOf?: number;
  maximum?: number;
  exclusiveMaximum?: number;
  minimum?: number;
  exclusiveMinimum?: number;

  /**
   * @see https://tools.ietf.org/html/draft-handrews-json-schema-validation-01#section-6.3
   */
  maxLength?: number;
  minLength?: number;
  pattern?: string;

  /**
   * @see https://tools.ietf.org/html/draft-handrews-json-schema-validation-01#section-6.4
   */

  additionalItems?: JSONSchema;
  maxItems?: number;
  minItems?: number;
  uniqueItems?: boolean;
  contains?: JSONSchema;

  /**
   * @see https://tools.ietf.org/html/draft-handrews-json-schema-validation-01#section-6.5
   */
  maxProperties?: number;
  minProperties?: number;

  patternProperties?: {
    [key: string]: JSONSchema;
  };
  additionalProperties?: JSONSchema | boolean;
  dependencies?: {
    [key: string]: JSONSchema | string[];
  };
  propertyNames?: JSONSchema;
  errorMessage?: ErrorMessage;

  /**
   * @see https://tools.ietf.org/html/draft-handrews-json-schema-validation-01#section-6.6
   */
  if?: JSONSchema;
  then?: JSONSchema;
  else?: JSONSchema;

  /**
   * @see https://tools.ietf.org/html/draft-handrews-json-schema-validation-01#section-7
   */
  format?: string;

  /**
   * @see https://tools.ietf.org/html/draft-handrews-json-schema-validation-01#section-8
   */
  contentMediaType?: string;
  contentEncoding?: string;

  /**
   * @see https://tools.ietf.org/html/draft-handrews-json-schema-validation-01#section-10
   */
  title?: string;
  description?: string;
  default?: JSONSchema7Type;
  readOnly?: boolean;
  writeOnly?: boolean;
  examples?: JSONSchema7Type;
}

export class JSONSchema extends JSONSchemaBase {
  uid?: string;
  properties?: {
    [key: string]: JSONSchema;
  };
  allOf?: JSONSchema[];
  anyOf?: JSONSchema[];
  oneOf?: JSONSchema[];
  not?: JSONSchema;

  /**
   * @see https://tools.ietf.org/html/draft-handrews-json-schema-validation-01#section-9
   */
  definitions?: {
    [key: string]: JSONSchema;
  };
  items?: JSONSchema;
  required?: string[];
  reference?: string;
}
