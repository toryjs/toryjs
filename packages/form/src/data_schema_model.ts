import * as validations from './validation';

// @ts-ignore
import initCustomErrors from 'ajv-errors';
import Ajv from 'ajv';

import { getRoot, getPath, IAnyStateTreeNode } from 'mobx-state-tree';

import { JSONSchema, JSONSchemaBase } from './json_schema';
import { safeEval } from './form_utils';
import { DataSet } from './form_store';

const defaultAjv = new Ajv({ allErrors: true, useDefaults: true, jsonPointers: true });
initCustomErrors(defaultAjv);

defaultAjv.addKeyword('validationExpression', {
  // type: 'string',]
  errors: true,
  validate: function v(
    this: any,
    schema: any,
    data: any,
    elementSchema: JSONSchema,
    elementPath: string,
    parentData: any
  ) {
    // console.log(arguments);
    // console.log(parentData);
    // console.log(schema);
    // console.log(data);
    // console.log(safeEval(parentData, schema, data));
    // console.log(elementPath);

    const result = safeEval(parentData, schema, data);
    if (!result) {
      // console.log(`Validation Error "${elementSchema.validationMessage}" at: ` + elementPath);

      (v as any).errors = [];
      (v as any).errors.push({
        keyword: 'validationExpression',
        dataPath: elementPath,
        message: elementSchema.errorMessage ? elementSchema.errorMessage : 'Value is invalid',
        params: {
          keyword: 'validationExpression'
        }
      });
    }
    return result;
  }
});

export type ListItem = {
  text: string;
  value: string;
};

type SchemaOptions = {
  parent?: Schema;
  required?: boolean;
  key?: string;
  ajv?: Ajv.Ajv;
  definitions?: any;
};

export class Schema extends JSONSchemaBase {
  // store: typeof FormStore.Type;
  originalSchema: JSONSchema;
  parent: Schema;
  properties: { [index: string]: Schema };
  definitions: { [index: string]: Schema };
  items: Schema;
  required: boolean;
  readOnly: boolean;
  enum: ListItem[];
  key: string;
  default: any;
  anyOf: JSONSchema[];

  validator: Ajv.ValidateFunction;

  constructor(
    input: JSONSchema,
    { parent = null, required = false, key = null, definitions = null }: SchemaOptions = {}
  ) {
    super();

    Object.assign(this, input);

    this.originalSchema = input;
    this.parent = parent;
    this.expression = input.expression;
    this.key = key;

    let schema = { ...input };

    if (definitions) {
      schema.definitions = { ...(schema.definitions || {}), ...definitions };
    }

    // we do not need validator for end nodes
    // these are always validated from the parent
    if (schema.properties) {
      this.validator = defaultAjv.compile(schema);
    }

    if (required) {
      this.required = required;
    }

    if (schema.type === 'object') {
      if (schema.properties) {
        this.properties = {};
        for (let key of Object.getOwnPropertyNames(schema.properties)) {
          this.properties[key] = new Schema(schema.properties[key] as JSONSchema, {
            parent: this,
            required: schema.required && schema.required.includes(key),
            key,
            definitions: schema.definitions
          });
        }
      }
    }

    if (schema.definitions) {
      this.definitions = {};
      for (let key of Object.getOwnPropertyNames(schema.definitions)) {
        this.definitions[key] = new Schema(schema.definitions[key] as JSONSchema, {
          parent: this,
          key
        });
      }
    }

    if (schema.type === 'array') {
      this.items = new Schema(schema.items as JSONSchema, {
        parent: this.parent,
        key: this.key,
        required: this.required,
        definitions: schema.definitions
      });
    }
  }

  rootSchema() {
    let s: Schema = this;
    while (s.parent != null) {
      s = s.parent;
    }
    return s;
  }

  // randomValue() {
  //   switch (this.type) {
  //     case 'string':
  //       return random.words(2);
  //     case 'boolean':
  //       return random.boolean();
  //     case 'id':
  //       return '1';
  //     case 'integer':
  //       return random.int();
  //     case 'number':
  //       return random.float();
  //   }
  //   return undefined;
  // }

  init<T>(schema: T, keys: Array<keyof Schema>) {
    for (let key of keys) {
      if ((schema as any)[key] != null) {
        this[key] = (schema as any)[key];
      }
    }
  }

  defaultValue<T>(dataset: T = {} as T): T {
    // validator is set to assign default values
    if (this.validator) {
      this.validator(dataset);
      return dataset;
    }
    return '' as any;
  }

  /* =========================================================
      Validations
     ======================================================== */

  static convertPath(path: string) {
    path = path.replace(/\/(\d+)/g, '[$1]');
    path = path.replace(/\//g, '.');

    return path;
  }

  static reassignErrors(errors: any[]) {
    if (!Array.isArray(errors)) {
      return errors;
    }

    for (let i = errors.length - 1; i >= 0; i--) {
      let error = errors[i];

      // remove string warnings where we use DateTime
      // this is a very dirty solution but allows me to store dates
      // in the dataset and ignore validations errors over them

      // if (error.message === 'should be string') {
      //   let path = error.schemaPath.split('/').slice(1, -1);
      //   let s: any = schema;
      //   while (path.length) {
      //     s = s[path.shift()];
      //   }
      //   if (s.format === 'date-time') {
      //     errors.splice(i, 1);
      //   }
      // }

      // missing properties are propagated into properties themselves
      // rendering as "value is required"
      if (error.params && error.params.missingProperty) {
        error.dataPath += '/' + error.params.missingProperty;
        error.message = 'Value is required';
      } else if (error.keyword === 'errorMessage') {
        let required = error.params.errors.find((p: Ajv.ErrorObject) => p.keyword === 'required');
        if (required) {
          error.dataPath += '/' + required.params.missingProperty;
        }
      }
    }
    return errors.length ? errors : undefined;
  }

  static parseParent(dataPath: string) {
    dataPath = Schema.convertPath(dataPath);

    if (dataPath.indexOf('.') >= 0) {
      return {
        dataPath: dataPath.substring(0, dataPath.lastIndexOf('.')),
        property: dataPath.substring(dataPath.lastIndexOf('.') + 1)
      };
    }
    return { property: dataPath, dataPath: '' };
  }

  tryParse(value: any) {
    if (!value) {
      return value;
    }
    if (this.format === 'date-time') {
      let date = Date.parse(value);
      return isNaN(date) ? value : new Date(date);
    }
    switch (this.type) {
      case 'integer':
        return validations.IntValidator(value) ? value : parseInt(value, 10);
      case 'number':
        return validations.FloatValidator(value) ? value : parseFloat(value);
      case 'boolean':
        return value === true || value === 'true' || value === 'True'
          ? true
          : !value || value === false || value === 'false' || value === 'False'
            ? false
            : value;
      default:
        return value;
    }
  }

  assignErrors(dataset: DataSet, error: Ajv.ErrorObject) {
    let { property, dataPath } = Schema.parseParent(error.dataPath);
    const node = eval(`dataset${dataPath}`);

    if (!property) {
      property = 'ROOT';
    }

    if (!node.errors.get(property)) {
      node.errors.set(property, error.message);
    }
  }

  validate(dataset: DataSet) {
    const cleanData = dataset.toJS ? dataset.toJS({ replaceDates: true }) : dataset;

    // we use this when executing expressions
    // (this.ajv as any).currentData = cleanData;

    let validator = this.validator;

    if (!validator(cleanData) as any) {
      return Schema.reassignErrors(validator.errors);
    }
    return false;
  }

  validateAndAssignErrors(dataset: DataSet) {
    const cleanData = dataset.toJS({ replaceDates: true });
    const errors = this.validate(cleanData);
    if (errors && errors.length) {
      for (let error of errors) {
        this.assignErrors(dataset, error);
      }
    }
    return errors;
  }

  validateFields(value: IAnyStateTreeNode, keys: string[], assignErrors = false) {
    // validate the whole dataset from its root
    const root: DataSet = getRoot(value) as any;
    const errors = this.validate(root as DataSet);

    let result = undefined;
    for (let key of keys) {
      // mobx's 'getPath' generates path such as /prop/path/0
      // we need to convert it to the json schema type path such as.prop.path[0]
      // const path = Schema.convertPath(getPath(value)) + (key ? '.' + key : '');
      const path = getPath(value) + (key ? '/' + key : '');

      // locate the error in the resulting errors by its path
      if (errors && errors.length) {
        const error = errors.find(e => e.dataPath === path);
        if (error) {
          if (assignErrors) {
            this.assignErrors(root, error);
          }
          result = error.message;
        }
      }
    }
    return result;
  }
}
