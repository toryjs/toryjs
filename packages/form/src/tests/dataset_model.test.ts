import { autorun } from 'mobx';

import merge from 'deepmerge';
import { fake } from 'sinon';

import { Schema } from '../data_schema_model';
import { buildStore } from '../mst_builder';
import { create } from './data';
import { JSONSchema } from '../json_schema';

import { config } from '../config';

describe('Dataset', () => {
  let jsonSchema = (): JSONSchema => ({
    type: 'object',
    properties: {
      name: {
        type: 'string'
      },
      age: {
        type: 'integer',
        minimum: 0,
        maximum: 130,
        default: 0
      },
      salary: {
        type: 'number',
        minimum: 0,
        default: 300
      },
      married: {
        type: 'boolean',
        default: false
      },
      dateMarried: {
        type: 'string',
        format: 'date-time'
      },
      fatherAge: {
        type: 'integer',
        validationExpression: 'value > this.age + 18',
        errorMessage: 'Father age must be at least 18 years more then your age'
      },
      accounts: {
        type: 'array',
        uniqueItems: true,
        items: {
          type: 'object',
          properties: {
            number: {
              type: 'string',
              pattern: '\\d\\d\\d-\\d\\d\\d'
            },
            money: {
              type: 'number',
              validationExpression: 'this.money % 2 === 0', // allow only to add higher value,
              errorMessage: 'You can only put even value of money!'
            }
          },
          required: ['number']
        },

        minItems: 1,
        maxItems: 3
      },
      accountTotal: {
        type: 'number',
        expression: 'this.accounts.reduce((p, n) => n.money + p, 0)'
      },
      address: {
        type: 'object',
        properties: {
          street: {
            type: 'string',
            minLength: 5,
            maxLength: 30
          },
          number: {
            type: 'number',
            exclusiveMinimum: 0,
            exclusiveMaximum: 1000
          }
        }
      }
    },
    required: ['name']
  });

  let schema = new Schema(jsonSchema());

  it('creates a default value', () => {
    expect(schema.defaultValue()).toMatchSnapshot();

    expect(schema.properties.accounts.items.defaultValue()).toMatchSnapshot();
  });

  it('creates a new mst and validates values', () => {
    config.setDirty = fake();

    const mst = buildStore(schema);
    const data = mst.create({});
    let error = '';
    // check conversion

    /* =========================================================
        Integer / Number
       ======================================================== */

    // ok

    data.setValue('age', '30');
    expect(data.getValue('age')).to.equal(30);

    data.setValue('salary', '30.15');
    expect(data.getValue('salary')).to.equal(30.15);

    data.setValue('salary', '30');
    expect(data.getValue('salary')).to.equal(30);

    // // format

    data.setValue('age', '30.15');
    expect(data.getValue('age')).to.equal('30.15');
    expect(data.validateField('age')).to.equal('should be integer');

    data.setValue('age', 'Momo');
    expect(data.getValue('age')).to.equal('Momo');
    expect(data.validateField('age')).to.equal('should be integer');

    data.setValue('salary', '30.aa');
    expect(data.getValue('salary')).to.equal('30.aa');
    expect(data.validateField('salary')).to.equal('should be number');

    // minimum

    data.setValue('age', '0');
    expect(data.getValue('age')).to.equal(0);

    data.setValue('age', '-1');
    expect(data.getValue('age')).to.equal(-1);
    expect(data.validateField('age')).to.equal('should be >= 0');

    // // maximum

    data.setValue('age', '130');
    expect(data.getValue('age')).to.equal(130);

    data.setValue('age', '131');
    expect(data.getValue('age')).to.equal(131);
    expect(data.validateField('age')).to.equal('should be <= 130');

    // subselection ok 'a.b.c'

    data.setValue('address.number', '20');
    expect(data.getValue('address.number')).to.equal(20);

    // exclusive minimum

    data.setValue('address.number', '0');
    expect(data.getValue('address.number')).to.equal(0);
    expect(data.getValue('address').validateField('number')).to.equal('should be > 0');

    data.setValue('address.number', '1');
    expect(data.getValue('address.number')).to.equal(1);

    // exclusive maximum

    data.setValue('address.number', '1000');
    expect(data.getValue('address.number')).to.equal(1000);
    expect(data.getValue('address').validateField('number', '1000')).to.equal('should be < 1000');

    data.setValue('address.number', '999');
    expect(data.getValue('address.number')).to.equal(999);

    /* =========================================================
        String
       ======================================================== */

    // ok

    data.setValue('name', 'Bobo');
    expect(data.getValue('name')).to.equal('Bobo');

    // format ok

    data.addRow('accounts');
    data.getValue('accounts')[0].setValue('number', '234-234');

    // console.log(3);
    // console.log(data.getValue('accounts')[0]);

    expect(data.getValue('accounts')[0].number).to.equal('234-234');

    error = data.getValue('accounts')[0].validateField('number', '123-123');
    expect(error).to.be.undefined;

    // format error

    data.getValue('accounts')[0].setValue('number', '123d');
    expect(data.getValue('accounts')[0].number).to.equal('123d');
    error = data.getValue('accounts')[0].validateField('number');
    expect(error).to.equal('should match pattern "\\d\\d\\d-\\d\\d\\d"');

    data.getValue('accounts')[0].setValue('number', '');
    error = data.getValue('accounts')[0].validateField('number', '');
    expect(error).to.equal('Value is required');

    // minLength

    data.getValue('address').setValue('street', '123456');
    error = data.getValue('address').validateField('street', '123456');
    expect(error).to.be.undefined;

    data.getValue('address').setValue('street', '123');
    error = data.getValue('address').validateField('street');
    expect(error).to.equal('should NOT be shorter than 5 characters');

    // maxLength

    data.getValue('address').setValue('street', '1234567890123456789012345678901234567890');
    error = data.getValue('address').validateField('street');
    expect(error).to.equal('should NOT be longer than 30 characters');

    /* =========================================================
        Array
       ======================================================== */

    // min intems

    data.removeRow('accounts', 0);
    error = data.validateField('accounts');
    expect(error).to.equal('should NOT have fewer than 1 items');

    // max items

    data.addRow('accounts');
    data.addRow('accounts');
    data.addRow('accounts');
    data.addRow('accounts');

    error = data.validateField('accounts');
    expect(error).to.equal('should NOT have more than 3 items');

    data.removeRow('accounts', 3);

    // unique items

    // console.log(toJS(data.getValue('accounts')));
    // console.log(data.getValue('accounts'))

    error = data.validateField('accounts');
    expect(error).to.equal('should NOT have duplicate items (items ## 1 and 2 are identical)');

    data.getValue('accounts')[1].setValue('number', '123-234');

    error = data.validateField('accounts');
    expect(error).to.equal('should NOT have duplicate items (items ## 0 and 2 are identical)');

    // all is well

    data.getValue('accounts')[0].setValue('number', '123-456');
    data.getValue('accounts')[2].setValue('number', '456-789');

    error = data.validateField('accounts');
    expect(error).to.be.undefined;

    // unique items

    /* =========================================================
        Expression
       ======================================================== */

    data.setValue('age', '10');
    data.setValue('fatherAge', '40');
    expect(data.getValue('fatherAge')).to.equal(40);

    // failing expression

    data.setValue('fatherAge', '20');
    expect(data.getValue('fatherAge')).to.equal(20);

    // console.log('WILL VALIDATE FATHER AGE');
    // console.log(data.fatherAge);

    error = data.validateField('fatherAge');
    expect(error).to.equal('Father age must be at least 18 years more then your age');

    data.setValue('fatherAge', '80');
    expect(data.getValue('fatherAge')).to.equal(80);
    error = data.validateField('fatherAge');
    expect(error).to.be.undefined;

    data.getValue('accounts')[0].setValue('money', 10);
    data.getValue('accounts')[0].setValue('money', 5);
    error = data.getValue('accounts')[0].validateField('money');
    expect(error).to.equal('You can only put even value of money!');
    expect(data.getValue('accounts')[0].errors.get('money')).to.equal(
      'You can only put even value of money!'
    );

    data.getSchema('fatherAge').errorMessage = null;
  });

  it('creates mst with values', () => {
    const mst = buildStore(schema);
    const data = mst.create({
      age: 15,
      name: 'Tomas',
      fatherAge: 50,
      married: true,
      dateMarried: create.date(),
      salary: 2300.34,
      address: {
        street: 'Elm street',
        number: 4
      },
      accounts: [{ number: '111-222', money: 200 }]
    });

    const errors = data.validateDataset();
    expect(errors).to.be.false;

    expect(data.toJS({ replaceDates: false })).toMatchSnapshot();
  });

  it('creates mst with default values', () => {
    const mst = buildStore(schema);
    const defaultData = mst.create({});
    expect(defaultData.toJS({ replaceDates: false, replaceEmpty: false })).toMatchSnapshot();
  });

  it('creates mst with external schemas', () => {
    const n = buildStore(
      {
        type: 'object',
        properties: {
          moo: { $ref: '#/definitions/moo' }
        }
      },
      {
        moo: {
          type: 'object',
          properties: {
            pip: { type: 'string' }
          }
        }
      }
    );
    const data = n.create({ moo: { pip: '1' } });
  });

  it('simulates polymorphism via merging', () => {
    const BaseElement = {
      type: 'object',
      properties: {
        id: { type: 'number' },
        name: { type: 'string' },
        documentation: { type: 'string' }
      }
    };

    // Flow has nothing, but sequence flow has

    const SequenceFlow = merge(BaseElement, {
      type: 'object',
      properties: {
        conditionExpression: { type: 'string' },
        id: { type: 'string' }
      }
    });

    console.log(SequenceFlow);
    expect(SequenceFlow.properties.id.type).to.equal('string');
    expect(SequenceFlow.properties.conditionExpression.type).to.equal('string');
  });

  it('creates mst with recursive definitions values', () => {
    const recursiveSchema: JSONSchema = {
      type: 'object',
      properties: {
        name: { type: 'string' },
        father: { $ref: '#' },
        friends: {
          type: 'array',
          items: { $ref: '#' }
        }
      }
    };
    const mst = buildStore(new Schema(recursiveSchema));
    const defaultData = mst.create({
      name: 'Tomas',
      father: { name: 'Michal Jr', father: { name: 'Michal Sr' } },
      friends: [{ name: 'Tomas' }, { name: 'Harry' }]
    });
    expect(defaultData.toJS({ replaceDates: false, replaceEmpty: false })).to.deep.equal({
      father: {
        father: { father: undefined, friends: [], name: 'Michal Sr' },
        friends: [],
        name: 'Michal Jr'
      },
      friends: [
        { father: undefined, friends: [], name: 'Tomas' },
        { father: undefined, friends: [], name: 'Harry' }
      ],
      name: 'Tomas'
    });
  });

  it('creates mst with referenced definitions values', () => {
    const recursiveSchema: JSONSchema = {
      type: 'object',
      properties: {
        name: { type: 'string' },
        car: { $ref: '#/definitions/car' },
        father: { $ref: '#' }
      },
      definitions: {
        car: {
          type: 'object',
          properties: {
            brand: { type: 'string' }
          }
        }
      }
    };
    const mst = buildStore(new Schema(recursiveSchema));
    const defaultData = mst.create({
      name: 'Tomas',
      father: { name: 'Michal Jr' },
      car: { brand: 'Honda' }
    });
    expect(defaultData.toJS()).to.deep.equal({
      car: { brand: 'Honda' },
      father: { car: undefined, father: undefined, name: 'Michal Jr' },
      name: 'Tomas'
    });
  });

  it('validates the root dataset', () => {
    const jSchema = jsonSchema();
    jSchema.oneOf = [{ required: ['name'] }, { required: ['age'] }];
    const rSchema = new Schema(jSchema);

    const store = buildStore(rSchema);
    const d1 = store.create({ name: 'Tomas', age: 39 });
    const errors = d1.validateDataset(); /*?*/
    expect(d1.errors.get('ROOT')).to.equal('should match exactly one schema in oneOf');
  });

  it('correctly removes "possible" invalid values on successful validation of the whole dataset', () => {
    config.setDirty = fake();
    const jSchema: JSONSchema = {
      type: 'object',
      properties: { foo: { type: 'string' }, bar: { type: 'string' } }
    };
    jSchema.anyOf = [{ required: ['foo'] }, { required: ['bar'] }];
    const rSchema = new Schema(jSchema);
    const store = buildStore(rSchema);
    let d1 = store.create({});
    d1.validateDataset(); /*?*/

    expect(d1.errors.get('foo')).to.equal('Value is required');
    expect(d1.errors.get('bar')).to.equal('Value is required');

    d1.setValue('foo', 'foo');
    d1.validateDataset(); /*?*/

    expect(d1.errors.get('foo')).to.be.undefined;
    expect(d1.errors.get('bar')).to.be.undefined;

    d1.setValue('foo', '');
    d1.setValue('bar', 'bar');
    d1.validateDataset(); /*?*/

    expect(d1.errors.get('foo')).to.be.undefined;
    expect(d1.errors.get('bar')).to.be.undefined;
  });

  it('respects validationGroup and correctly removes "possible" invalid values on successful validation', () => {
    config.setDirty = fake();

    const jSchema: JSONSchema = {
      type: 'object',
      properties: {
        foo: { type: 'string', validationGroup: '1' },
        bar: { type: 'string', validationGroup: '1' },
        dar: { type: 'string' }
      }
    };
    jSchema.anyOf = [{ required: ['foo'] }, { required: ['bar'] }];
    const rSchema = new Schema(jSchema);
    const store = buildStore(rSchema);
    let d1 = store.create({});

    d1.validateDataset();
    expect(d1.errors.get('foo')).to.equal('Value is required');
    expect(d1.errors.get('bar')).to.equal('Value is required');

    d1.setValue('foo', 'foo');

    expect(d1.errors.get('foo')).to.equal('');
    expect(d1.errors.get('bar')).to.equal('');

    d1.setValue('foo', '');

    expect(d1.errors.get('foo')).to.equal('Value is required');
    expect(d1.errors.get('bar')).to.equal('Value is required');
  });

  it('allows to use expressions', () => {
    const mst = buildStore(schema);
    const defaultData = mst.create({
      accounts: [{ number: '1', money: 1 }, { number: '2', money: 2 }, { number: '3', money: 3 }]
    });

    expect(defaultData.getValue('accountTotal')).to.equal(6);

    let finalAccount = 0;
    autorun(() => {
      finalAccount = defaultData.getValue('accountTotal');
    });

    // check computed fields
    defaultData.getValue('accounts')[0].setValue('money', 10);
    expect(finalAccount).to.equal(15);
  });
});
