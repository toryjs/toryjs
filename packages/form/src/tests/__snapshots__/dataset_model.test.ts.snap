// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dataset creates a default value 1`] = `
Object {
  "age": 0,
  "married": false,
  "salary": 300,
}
`;

exports[`Dataset creates a default value 2`] = `Object {}`;

exports[`Dataset creates mst with default values 1`] = `
Object {
  "accounts": Array [],
  "address": Object {
    "number": "",
    "street": "",
  },
  "age": "",
  "dateMarried": "",
  "fatherAge": "",
  "married": "",
  "name": "",
  "salary": 300,
}
`;

exports[`Dataset creates mst with values 1`] = `
Object {
  "accounts": Array [
    Object {
      "money": 200,
      "number": "111-222",
    },
  ],
  "address": Object {
    "number": 4,
    "street": "Elm street",
  },
  "age": 15,
  "dateMarried": 2017-02-02T00:00:00.000Z,
  "fatherAge": 50,
  "married": true,
  "name": "Tomas",
  "salary": 2300.34,
}
`;
